walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,66 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import com.google.sps.Event;
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Set<String> allAttendees = new HashSet<>(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+    
+    Collection<TimeRange> ranges = getTimeRanges(events, allAttendees, request.getDuration());
```

Comment:
generally speaking, we want to prefer code with less branching, and less return points. Can this be rewritten to minimize the number of return points?

Why do we prefer this? Because less branching and less return points means less complexity to the reader. There is one notable exception to the return points idea - early exit for exceptional conditions. For example, I would prefer:

void doSomething() {
  if (errorCondition) {
    return;
  }

  ...
}

instead of:

void doSomething() {
  if (!errorCondition) {
    ...
  }
}

So there's no hard and fast rule, but we do want to consider what is easiest for others to read and comprehend.

Reply:
Refactored to reduce number of returns.